### PR TITLE
support SMAC as a tuner on nni

### PR DESCRIPTION
**Pull request by QuanluZhang**
1. support SMAC, (SMAC package should be installed separately)
2. fix the bug in sdk logging, i.e., endless logging
3. fix bug in nnictl launcher
4. fix bug in setup.py, i.e., incorrectly adding another entry_point

### CODE REVIEW

File: src/sdk/pynni/nni/__main__.py [hunk 1]
```diff
@@ -93,19 +105,28 @@ def main():
     tuner = None
     assessor = None
 
-    if args.tuner_class_name is None:
-        raise ValueError('Tuner must be specified')
-    if args.tuner_class_name in BUILT_IN_CLASS_NAMES:
-        tuner = create_builtin_class_instance(args.tuner_class_name, args.tuner_args)
+    if args.tuner_class_name in ModuleName:
```
[the remark this answers is not in the document]
Comment (QuanluZhang) [reply src/sdk/pynni/nni/__main__.py]: this is guaranteed by nnictl, also argparse makes tuner_class_name required.

File: src/sdk/pynni/nni/__main__.py [hunk 2]
```diff
@@ -93,19 +105,28 @@ def main():
...
         else:
-            assessor = create_customized_class_instance(args.assessor_directory, \
-                args.assessor_class_filename, args.assessor_class_name, args.assessor_args)
+            assessor = create_customized_class_instance(
+                args.assessor_directory,
+                args.assessor_class_filename,
+                args.assessor_class_name,
+                args.assessor_args)
 
     if tuner is None:
```
[the remark this answers is not in the document]
Comment (QuanluZhang) [reply src/sdk/pynni/nni/__main__.py]: I can move it to line 118, agree?

File: src/sdk/pynni/nni/smac_tuner/convert_ss_to_scenario.py [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+    # https://automl.github.io/SMAC3/stable/options.html
+    search_space = nni_search_space_content
+    with open('param_config_space.pcs', 'w') as pcs_fd:
+        if isinstance(search_space, dict):
+            for key in search_space.keys():
+                if isinstance(search_space[key], dict):
+                    try:
+                        if search_space[key]['_type'] == 'choice':
+                            pcs_fd.write('%s categorical {%s} [%s]\n' % (
+                                key,
```
[the remark this answers is not in the document]
Comment (QuanluZhang) [reply src/sdk/pynni/nni/smac_tuner/convert_ss_to_scenario.py]: where? do you means the comments from line 34 to line 39?

File: src/sdk/pynni/nni/smac_tuner/smac_tuner.py [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+        else:
+            optimizer = None
+
+        return optimizer
+
+    def update_search_space(self, search_space):
+        '''
+        TODO: this is urgly, we put all the initialization work in this method, 
+        because initialization relies on search space, also because update_search_space is called at the beginning.
+        NOTE: updating search space is not supported.
```
[the remark this answers is not in the document]
Comment (QuanluZhang) [reply src/sdk/pynni/nni/smac_tuner/smac_tuner.py]: this means user cannot update search space.

File: src/sdk/pynni/nni/smac_tuner/smac_tuner.py [hunk 5]
```diff
@@ -0,0 +1,190 @@
...
+            self.update_ss_done = True
+        else:
+            pass
+
+    def receive_trial_result(self, parameter_id, parameters, reward):
+        '''
+        receive_trial_result
+        '''
+        if self.optimize_mode is OptimizeMode.Maximize:
+            reward = -reward
```
[the remark this answers is not in the document]
Comment (QuanluZhang) [reply src/sdk/pynni/nni/smac_tuner/smac_tuner.py]: They optimize cost.

File: src/sdk/pynni/nni/smac_tuner/convert_ss_to_scenario.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (xuehui1991) [reply src/sdk/pynni/nni/smac_tuner/convert_ss_to_scenario.py]: I means we should mention it in README

File: src/sdk/pynni/nni/smac_tuner/smac_tuner.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (xuehui1991) [reply src/sdk/pynni/nni/smac_tuner/smac_tuner.py]: But when init tuner, we could call this function, right?

File: src/sdk/pynni/nni/smac_tuner/convert_ss_to_scenario.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (QuanluZhang) [reply src/sdk/pynni/nni/smac_tuner/convert_ss_to_scenario.py]: sure, i will write a detailed README for the smac tuner to demonstrate how the original smac/configspace repo is modified and how smac is integrated on nni. Also, I will update our current doc to show which prior distributions the smac tuner supports.

File: src/sdk/pynni/nni/smac_tuner/smac_tuner.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (QuanluZhang) [reply src/sdk/pynni/nni/smac_tuner/smac_tuner.py]: This implementation is ugly. Init smac tuner requires search space, however search space is provided only when update_search_space is called. This makes me move initialization action into the update_search_space function. If I do init things in __init__ method, user has to provide search space file again in `classArgs` which is duplicated with `searchSpacePath`. For your question, the update_search_space method is called when tuner is launched.

**Comment by yds05** [comment]
Make please sure you use squash merge after the review is completed.

**Pull request merged**